pkg/ottl/ottlfuncs/func_delete_key.go

```diff
@@ -16,15 +16,39 @@ package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-c
 
 import (
 	"context"
+	"fmt"
 
 	"go.opentelemetry.io/collector/pdata/pcommon"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
 )
 
-func DeleteKey[K any](target ottl.Getter[K], key string) (ottl.ExprFunc[K], error) {
+type DeleteKeyArguments[K any] struct {
+	Target ottl.Getter[K] `ottlarg:"0"`
+	Key    string         `ottlarg:"1"`
+}
+
+type DeleteKeyFactory[K any] struct{}
+
+func (f DeleteKeyFactory[K]) Name() string { return "delete_key" }
+
+func (f DeleteKeyFactory[K]) CreateEmptyArguments() ottl.Arguments {
+	return &DeleteKeyArguments[K]{}
+}
+
+func (f DeleteKeyFactory[K]) CreateFunction(ctx context.Context, fCtx ottl.FunctionContext, args ottl.Arguments) (ottl.ExprFunc[K], error) {
+	dkargs, ok := args.(*DeleteKeyArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("DeleteKeyFactory args must be of type *DeleteKeyArguments[K]")
+	}
+
+	return deleteKey(ctx, fCtx, dkargs), nil
+}
+
+func deleteKey[K any](_ context.Context, fCtx ottl.FunctionContext, args *DeleteKeyArguments[K]) ottl.ExprFunc[K] {
```

Comment:
Does the `context.Context` argument buy us anything in this function?

Reply:
There are no use cases yet that require context.Context in the function creation (or the functions themselves technically)

Reply:
I would say definitely more important potential for in the ExprFunc since they handle hot path execution.  For the function creation during startup it feels less important, but is a Go best practice to pass it everywhere right?

Reply:
I'm mainly just thinking for the `deleteKey` implementation function, it seems like it won't be helpful since it's just intended to be a simple closure, and I don't think we'd want the function creation context available in the ExprFunc.

Looking at the [context package documentation](https://pkg.go.dev/context#pkg-overview) however, I'm also having trouble justifying it in `CreateFunction`. The package description says:
>Package context defines the Context type, which carries deadlines, cancellation signals, and other request-scoped values across API boundaries and between processes. 

Right now function creation is single-threaded, fast, and is only done on component initialization. Unless we plan to change one of these aspects, it's not clear to me how we will be able to effectively make use of `context.Context`. It may be considered a best practice, but the closest thing I can find claiming that comes from the conclusion of the [context package announcement](https://go.dev/blog/context), which still claims it is for request-scoped functionality.

I don't mind including it in `CreateFunction`, but it would be nice if we had a use in mind.

Reply:
> I don't think we'd want the function creation context available in the ExprFunc.

Definitely agree.  I am good not including it in CreateFunction since we have no valid use cases and very good with not passing it to `deleteKey` since during hot path execution the function should be using the context passed to it, not the one present during function creation.

Reply:
I've removed `context.Context` from the interface for now. If we come up with a use while we are finishing this up, we can add it back. I still think it is unlikely to see much use if it were included, but I will look at the component initialization process again to understand if there's a nuance I missed.

@bogdandrutu I'd also like to ask for your input here since your proposal for the interface included `context.Context`. Do you think `context.Context` would still be helpful in function creation given our discussion above?

Reply:
After looking into this some more, I think the best place to add `context.Context` would likely be in `ParseStatements`, which is the only part of the code I think could be slow enough to be considered blocking. I still can't see any other obvious uses for this during statement parsing. A `context.Context` object will still be passed into the `ExprFunc` if we want to do anything with it while processing data, which I think is more likely to see usage.

Reply:
I agree.